test/cpp/jit/test_gpu.cpp

```diff
@@ -426,6 +426,136 @@ void testGPU_FusionClear() {
   TORCH_CHECK(output_ref.equal(output));
 }
 
+void testGPU_FusionCopy() {
+  Fusion original_fusion;
+
+  // Create the test IR
+  {
+    FusionGuard fg(&original_fusion);
+
+    auto tv0 = makeDummyTensor(3);
+    auto tv1 = makeDummyTensor(3);
+    auto tv2 = add(tv1, new Float(2.0));
+    auto tv3 = sub(add(tv0, mul(tv2, tv2)), tv2);
+
+    original_fusion.addInput(tv0);
+    original_fusion.addInput(tv1);
+    original_fusion.addOutput(tv3);
+
+    tv3->reorder({{0, 2}, {2, 0}});
+    tv3->split(-1, 4);
+    tv3->reorder({{2, 0}, {3, 1}, {0, 3}});
+
+    tv0->computeAt(tv3, -1);
+    tv1->computeAt(tv3, -1);
+
+    tv3->axis(0)->parallelize(ParallelType::BIDx);
+    tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  }
+
+  // Test copy before lowering
+  Fusion clone = original_fusion;
+
+  // Compare IR dumps
+  std::stringstream original_ir;
+  std::stringstream clone_ir;
+  original_ir << original_fusion;
+  clone_ir << clone;
+  ASSERT_EQ(original_ir.str(), clone_ir.str());
+
+  // Lower original fusion
+  std::stringstream original_kernel;
+  {
+    GPULower lower(&original_fusion);
+    lower.printKernel(original_kernel);
+  }
+
+  // Make sure the "before lowering" clone was not mutated
+  // while lowering the original fusion IR
+  std::stringstream before_lowering_ir;
+  before_lowering_ir << clone;
+  ASSERT_EQ(original_ir.str(), before_lowering_ir.str());
+
+  // Test copy after lowering (including assignment operator)
+  Fusion before_lowering = clone;
+  clone = original_fusion;
+
+  // Compare IR dumps
+  std::stringstream original_lowered_ir;
+  std::stringstream clone_lowered_ir;
+  original_lowered_ir << original_fusion;
+  clone_lowered_ir << clone;
+  ASSERT_EQ(original_lowered_ir.str(), clone_lowered_ir.str());
+
+  // Lower the "before lowering" and compare kernels
+  std::stringstream clone_kernel;
+  {
+    GPULower lower(&before_lowering);
+    lower.printKernel(clone_kernel);
+  }
+  ASSERT_EQ(original_kernel.str(), clone_kernel.str());
+}
+
+void testGPU_FusionMove() {
+  Fusion fusion;
+
+  // Create the test IR
+  {
+    FusionGuard fg(&fusion);
+
+    auto tv0 = makeDummyTensor(3);
+    auto tv1 = makeDummyTensor(3);
+    auto tv2 = add(tv1, new Float(2.0));
+    auto tv3 = sub(add(tv0, mul(tv2, tv2)), tv2);
+
+    fusion.addInput(tv0);
+    fusion.addInput(tv1);
+    fusion.addOutput(tv3);
+
+    tv3->reorder({{0, 2}, {2, 0}});
+    tv3->split(-1, 4);
+    tv3->reorder({{2, 0}, {3, 1}, {0, 3}});
+
+    tv0->computeAt(tv3, -1);
+    tv1->computeAt(tv3, -1);
+
+    tv3->axis(0)->parallelize(ParallelType::BIDx);
+    tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  }
+
+  std::stringstream original_ir;
+  original_ir << fusion;
+
+  // Test move before lowering
+  Fusion another_fusion = std::move(fusion);
```

Comment:
nitpick: should we check move constructor does properly empty/clear the original?

Reply:
Good question! Yes, it's very tempted to do so, although generally the moved-away from objects are in a valid but unspecified state (for ex. in the standard library you can only call methods w/o preconditions, ex. clear() or assignment operators) - search for "preconditions" here: https://en.cppreference.com/w/cpp/utility/move

For this reason I decided against checking for it, but at the same time it's easy to argue that test cases may afford to make extra assumptions. So I'm going ahead and adding these checks, thanks for bringing it up.